### ISSUE TITLE
[Performance] pl.matmul(b_trans=True) costs 16-23% at shallow inner-K tiles despite identical generated code

### ISSUE THREAD

**Issue by lyfne123** [hw-native-sys/pypto#2309]
### Summary

`pl.matmul(..., b_trans=True)` is presented (and was adopted in `models/qwen3/14b`)
as a free way to consume a caller's `[out, in]` weight layout: it reinterprets the
tile rather than moving data, and the generated code confirms that — the two forms
differ only by a `blayout`/`slayout` swap on one `alloc_tile`, with no `ttrans`, no
extra load, and no extra instruction.

At runtime it is **not** free, and the cost depends sharply on the inner-K tile depth:

| inner tile (TK, TN) | `[K,N]` plain | `[N,K]` + `b_trans` | overhead |
| --- | --- | --- | --- |
| 256, 256 | 3.039 ms | 3.106 ms | **+2.2 %** |
| 64, 512  | 3.117 ms | 3.610 ms | **+15.8 %** |
| 64, 1024 | 6.825 ms | 8.406 ms | **+23.2 %** |

(40-layer stack, M=16, K=5120, medians of 10 dispatches; the repeat-to-repeat noise
floor of this harness measured ±1.4 %, so the 256/256 row is at noise and the TK=64
rows are far outside it.)

TK=64 is not a corner case: in `models/qwen3/14b/decode_fwd.py` it is the inner-K of
`out_proj` (64, 512) and of `gate`/`up`/`down` (64, 1024) — together **89 % of the
decode weight bytes**. Only `qkv` uses (256, 256).

Related: #1776

### Git Commit ID

4b0c2275945ea7aedf1c80ab5bd76e32db4791a8

### NPU Kind

Ascend 910B

### Host Platform

Linux (aarch64)

### Reproduction Script

```python
"""Minimal repro: `pl.matmul(..., b_trans=True)` costs 16-23% at shallow inner-K.

Two kernels compute the same y = x @ W. One takes W as [K, N] and multiplies
normally; the other takes the SAME numbers as [N, K] and sets b_trans=True.

The generated code differs only in the weight tile's declared layout --
`blayout`/`slayout` swap from col_major/row_major to row_major/col_major, an
NZ<->ZN fractal reinterpretation. There is no `ttrans`, no extra load, no extra
instruction: diffing the two `.pto` files (after normalising SSA names with
`sed -E 's/inline[0-9]+/inlineN/g'`) shows only the `alloc_tile` type line.
That is why b_trans was adopted as a free way to consume a caller's [out, in]
weight layout. It is not free at runtime.

**The overhead is a function of the inner-K tile depth**, so the tile shape must
be swept -- a repro pinned at one shape will miss it entirely:

    REPRO_TK=256 REPRO_TN=256  REPRO_N=5120   ->  +2.2%   (qkv)
    REPRO_TK=64  REPRO_TN=512  REPRO_N=5120   -> +15.8%   (out_proj)
    REPRO_TK=64  REPRO_TN=1024 REPRO_N=17408  -> +23.2%   (gate/up/down)

Two properties of the harness are deliberate, not padding. A single
[16,5120]x[5120,5120] matmul takes ~0.05 ms of device time against ~1.3 ms of
fixed dispatch overhead, so the effect drowns; and re-reading one small weight
tile would sit in cache, whereas the real decode kernel streams ~26 GB of
weights per step. Walking a 40-layer stack (2.0-6.6 GiB) restores both: device
time dominates, and the weights are actually streamed from GM.

Usage (on device):
    python btrans_repro.py -d 0 --iters 10
"""

import argparse
import statistics
import time

import pypto.language as pl
import torch
from pypto.backend import BackendType, set_backend_type
from pypto.runtime import ChipWorker, RunConfig

import os

# The real decode kernel does NOT use one tile shape. qkv is (TK, TN) = (256, 256)
# but out_proj is (64, 512) and gate/up/down are (64, 1024) -- and those three
# carry most of the weight bytes, so a repro pinned at (256, 256) exercises the
# smallest share of the work. Sweep them.
LAYERS = int(os.environ.get("REPRO_LAYERS", "40"))   # stack so the weights stream
M, K = 16, 5120                                      # batch x hidden
N = int(os.environ.get("REPRO_N", "17408"))
TK = int(os.environ.get("REPRO_TK", "256"))
TN = int(os.environ.get("REPRO_TN", "256"))
N_BLOCKS = N // TN          # one SPMD block per output tile
K_CHUNKS = K // TK


@pl.jit
def mm_plain(
    x: pl.Tensor[[M, K], pl.BF16],
    w: pl.Tensor[[LAYERS * K, N], pl.BF16],
    out: pl.Out[pl.Tensor[[M, N], pl.FP32]],
):
    """W is [LAYERS*K, N] -- PyPTO's own orientation, layer-major by K."""
    for layer in pl.range(LAYERS):
        lb = layer * K
        with pl.manual_scope():
            with pl.spmd(N_BLOCKS, name_hint="mm_plain", allow_early_resolve=True):
                n0 = pl.tile.get_block_idx() * TN
                acc = pl.matmul(x[:, 0:TK], w[lb : lb + TK, n0 : n0 + TN], out_dtype=pl.FP32)
                for kc in pl.pipeline(1, K_CHUNKS, stage=2):
                    k0 = kc * TK
                    acc = pl.matmul_acc(
                        acc, x[:, k0 : k0 + TK], w[lb + k0 : lb + k0 + TK, n0 : n0 + TN]
                    )
                out = pl.assemble(out, acc, [0, n0])
    return out


@pl.jit
def mm_btrans(
    x: pl.Tensor[[M, K], pl.BF16],
    w: pl.Tensor[[LAYERS * N, K], pl.BF16],
    out: pl.Out[pl.Tensor[[M, N], pl.FP32]],
):
    """W is [LAYERS*N, K] -- the caller's orientation, layer-major by N."""
    for layer in pl.range(LAYERS):
        lb = layer * N
        with pl.manual_scope():
            with pl.spmd(N_BLOCKS, name_hint="mm_btrans", allow_early_resolve=True):
                n0 = pl.tile.get_block_idx() * TN
                acc = pl.matmul(
                    x[:, 0:TK], w[lb + n0 : lb + n0 + TN, 0:TK], out_dtype=pl.FP32, b_trans=True
                )
                for kc in pl.pipeline(1, K_CHUNKS, stage=2):
                    k0 = kc * TK
                    acc = pl.matmul_acc(
                        acc, x[:, k0 : k0 + TK], w[lb + n0 : lb + n0 + TN, k0 : k0 + TK],
                        b_trans=True,
                    )
                out = pl.assemble(out, acc, [0, n0])
    return out


def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("-p", "--platform", default="a2a3")
    ap.add_argument("-d", "--device", type=int, default=0)
    ap.add_argument("--iters", type=int, default=10)
    args = ap.parse_args()

    set_backend_type(BackendType.Ascend910B)
    cfg = RunConfig(platform=args.platform, device_id=args.device)

    g = torch.Generator().manual_seed(0)
    x = torch.empty(M, K).normal_(0.0, 1.0, generator=g).to(torch.bfloat16)
    w1 = torch.empty(K, N).normal_(0.0, 0.02, generator=g).to(torch.bfloat16)
    w = torch.cat([w1] * LAYERS, dim=0).contiguous()                       # [LAYERS*K, N]
    wt = torch.cat([w1.transpose(0, 1).contiguous()] * LAYERS, dim=0).contiguous()  # [LAYERS*N, K]
    golden = x.float() @ w1.float()              # every layer holds the same weights

    results = {}
    for label, fn, weight in (("[K,N] plain", mm_plain, w), ("[N,K] b_trans", mm_btrans, wt)):
        out_h = torch.zeros(M, N, dtype=torch.float32)
        # Upload once and time only the dispatch: calling the jit entry directly
        # would take the one-shot path and re-upload the weights every call,
        # turning this into an H2D bandwidth test.
        with ChipWorker(config=cfg, level=2) as worker:
            xd = worker.alloc_tensor(list(x.shape), x.dtype, init=x)
            wd = worker.alloc_tensor(list(weight.shape), weight.dtype, init=weight)
            od = worker.alloc_tensor(list(out_h.shape), out_h.dtype, init=out_h)
            handle = worker.register(fn.compile(x, weight, out_h, config=cfg))
            handle(xd, wd, od, config=cfg)                      # warm-up
            ts = []
            for _ in range(args.iters):
                t0 = time.perf_counter()
                handle(xd, wd, od, config=cfg)
                ts.append((time.perf_counter() - t0) * 1e3)
            # DeviceTensor.data_ptr is an attribute, not a method (torch's is a method).
            worker.copy_from(out_h.data_ptr(), od.data_ptr, out_h.numel() * 4)
        close = torch.isclose(out_h, golden, rtol=2e-2, atol=2e-2).float().mean().item()
        results[label] = statistics.median(ts)
        print(f"  {label:<16} median {results[label]:7.3f} ms  "
              f"(min {min(ts):.3f} max {max(ts):.3f})  vs torch: close {100*close:.2f}%")

    a, b = results["[K,N] plain"], results["[N,K] b_trans"]
    gib = LAYERS * K * N * 2 / 1024 ** 3
    print(f"\n  LAYERS={LAYERS} M={M} K={K} N={N} TK={TK} TN={TN} | streamed {gib:.1f} GiB")
    print(f"  b_trans overhead: {b - a:+.3f} ms ({100 * (b - a) / a:+.1f}%)")


if __name__ == "__main__":
    main()
```

Sweep the tile shapes the decode kernel actually uses:

```bash
REPRO_TK=256 REPRO_TN=256  REPRO_N=5120  python btrans_repro.py -d 0 --iters 10   # qkv
REPRO_TK=64  REPRO_TN=512  REPRO_N=5120  python btrans_repro.py -d 0 --iters 10   # out_proj
REPRO_TK=64  REPRO_TN=1024 REPRO_N=17408 python btrans_repro.py -d 0 --iters 10   # gate/up/down
```

Both variants validate against a torch golden (`close 100.00%`), so this is a pure
timing difference, not a different computation.

### Expected Performance

`b_trans=True` should cost approximately nothing relative to feeding the same numbers
pre-transposed as `[K, N]`. That is the premise on which it is usable as a layout
adapter: the tile is reinterpreted (NZ<->ZN fractal), no data is moved, and the
emitted instruction stream is identical.

Concretely: the `+2.2 %` measured at (256, 256) is what all tile shapes were expected
to look like.

### Actual Performance

`+15.8 %` at (64, 512) and `+23.2 %` at (64, 1024) — i.e. the overhead grows as the
inner-K tile gets shallower, which is exactly the regime the Qwen3-14B decode kernel
uses for the majority of its weight traffic.

Whole-kernel effect, same die, 40 layers, weights uploaded once and only the dispatch
timed (`decode_fwd_layers` vs `decode_layers_fwd`, which differ only in weight
addressing):

```
[in,out]           37.544 ms   (min 37.352 max 42.249)
[out,in]+b_trans   43.693 ms   (min 43.601 max 43.976)   +6.149 ms  (+16.4 %)
```

Weighting the per-shape overheads by weight bytes predicts the same figure:
`0.11 x 2.2 % + 0.89 x ~20 % ~= 18 %`.

### Profiling Data (Optional)

Normalised `.pto` diff of one `q_proj` scope, `[in,out]` vs `[out,in]`+`b_trans`
(SSA names normalised with `sed -E 's/inline[0-9]+/inlineN/g; s/%[0-9]+/%N/g'`). The
only structural change is an extra `alloc_tile` type declaration with the layout pair
swapped; every `pto.textract` keeps an identical **output** type:

```diff
+ %t__tile_t = pto.alloc_tile ... !pto.tile_buf<loc=mat, dtype=bf16, rows=256, cols=256,
+     blayout=row_major, slayout=col_major, fractal=512, pad=0>

-   pto.textract ins(%N, ... blayout=col_major, slayout=row_major, fractal=512 ...)
-       outs(... loc=right, rows=64, cols=256, blayout=row_major, slayout=col_major ...)
+   pto.textract ins(%t__tile_t, ... blayout=row_major, slayout=col_major, fractal=512 ...)
+       outs(... loc=right, rows=64, cols=256, blayout=row_major, slayout=col_major ...)
```

There is no `ttrans` anywhere in either file and the instruction count is unchanged
(+4 lines total, all `alloc_tile` type declarations).

An L2 swimlane of the full decode kernel shows the cube pipe at **90.3 %** occupancy
with only 1.8 % fully-idle time, so this is not scheduling: the extra time is spent
inside the cube tasks themselves.

### Additional Context

Hardware: Ascend 910 A3 (`npu-smi` reports `Ascend910`, PCI 0xD803), driver 26.0.rc1,
CANN 9.0.0, 24 AIC + 48 AIV per die. PyPTO platform `a2a3`, `BackendType.Ascend910B`.

Why this matters beyond the microbenchmark: the vLLM-Ascend bridge for the Qwen3-14B
decode mega kernel adopted `[out, in]` + `b_trans` specifically so the kernel and
vLLM could **share one weight buffer** instead of holding a transposed second copy
(~26 GB at 40 layers, which does not fit on a 64 GB die alongside a KV pool). The
memory win is real and it is what makes 40 layers runnable at all — but it currently
costs 6.1 ms per decode step, which turns a ~1 ms device-time win over native
ACL-graph into a ~5 ms loss.

If the shallow-K `textract` path can be brought to the (256, 256) level, both
properties are available at once.

**Comment by lyfne123** [comment]
## Root cause: it is not the reinterpretation, it is the `tload` that fills the tile

The original report's instruction-level evidence is correct — and that is exactly what made it
misleading. The reinterpretation really is free: `%t__tile_t` aliases the loaded tile at the same
address and costs zero instructions. The cost is one instruction **upstream** of it, in the
`pto.tload` that fills the L1 (`loc=mat`) weight tile. Its shape is transposed along with the
layout, and with it the contiguous run per GM row:

```
plain  [K,N]:  %w_view = make_tensor_view ..., shape = [204800, 17408], strides = [17408, 1]
               %pview  = partition_view %w_view, offsets = [..], sizes = [64, 1024]
               pto.tload ins(%pview) outs(%0 : tile_buf<loc=mat, bf16, rows=64,   cols=1024, ...>)

b_trans[N,K]:  %w_view = make_tensor_view ..., shape = [696320,  5120], strides = [ 5120, 1]
               %pview  = partition_view %w_view, offsets = [..], sizes = [1024, 64]
               pto.tload ins(%pview) outs(%0 : tile_buf<loc=mat, bf16, rows=1024, cols=64,   ...>)
```

| | loaded tile | source row length | contiguous run | descriptors |
| --- | --- | --- | --- | --- |
| `[K,N]` plain | `[64, 1024]` | 17408 | **2048 B** | 64 |
| `[N,K]` b_trans | `[1024, 64]` | 5120 | **128 B** | 1024 |

Same 128 KB moved, 16x the descriptors, 1/16 the burst length. That is why the overhead tracks the
inner-K tile depth and nothing else, and why `qkv` escaped — it already loads (256, 256), i.e. a
512 B run either way.

## Separating "layout" from "loaded shape"

The above predicts the cost belongs to the loaded shape, not to which layout produced it. Two extra
variants test that directly: same task decomposition, same tile bytes, same MAC count, but the
inner loop re-split into 256-wide N sub-tiles with a 256-deep K chunk, so **both** layouts issue the
same (256, 256) `tload`.

```python
for j in pl.unroll(N_SUBS):                      # N_SUBS = TN // 256
    nj = n0 + j * 256
    acc = pl.matmul(x[:, 0:256], w[lb + nj : lb + nj + 256, 0:256],
                    out_dtype=pl.FP32, b_trans=True)
    for kc in pl.pipeline(1, K // 256, stage=2):
        k0 = kc * 256
        acc = pl.matmul_acc(acc, x[:, k0 : k0 + 256],
                            w[lb + nj : lb + nj + 256, k0 : k0 + 256], b_trans=True)
    out = pl.assemble(out, acc, [0, nj])
```

40-layer stack, M=16, K=5120, medians of 10 dispatches, all four variants in one process:

| shape | `[K,N]` plain | `[N,K]` b_trans | plain re-tiled | b_trans re-tiled |
| --- | --- | --- | --- | --- |
| out_proj, N=5120, (TK,TN)=(64,512) | 2.968 | 3.436 (+15.8%) | 2.940 | **3.069** |
| MLP, N=17408, (TK,TN)=(64,1024) | 6.696 | 8.287 (+23.7%) | 6.924 | **6.929** |

At an identical `tload` shape the two layouts measure **6.924 vs 6.929 ms — 0.07% apart**. The
layout is genuinely free; 100% of the penalty is the loaded shape. Note also that the residual
+3.5% of the re-tiled b_trans over plain@(64,1024) equals `plain re-tiled` over `plain` (+3.4%) —
that part is the narrower N tile, which the `[K,N]` layout pays too.

## Effect on the real kernel

Re-tiling the seven weight-matmul sites in the Qwen3-14B `[out, in]` decode layer body to
(256, 256) — task decomposition, dependency topology, tile bytes and MAC count all unchanged:

| | 40-layer dispatch | e2e device occupancy | fused kernel |
| --- | --- | --- | --- |
| before | 44.578 / 45.469 ms | 42.50 ms | 40.944 ms |
| after | **40.599 / 40.957 ms** | **38.88 ms** | **37.145 ms** |
| `[in, out]` reference | 40.168 ms | — | — |

A/B'd against the same binary in one session via an env switch, interleaved on/off/on/off, because
absolute times on this host drift >10% across sessions (an unmodified `[in, out]` baseline measured
32.829 ms in the morning and 40.168 ms in the evening — any before/after spanning two sessions is
drift plus effect). The two repeats of each arm agree within 1–2% while the arms differ by 10%.
Numerics unchanged: 100% close, and closer to the reference than the reference is to itself.

So the whole-kernel gap against `[in, out]` goes from +4.85 ms to **+0.61 ms (+1.5%)**.

## What this suggests upstream

The user-visible contract of `b_trans=True` is "consume a `[N, K]` operand". Nothing in it implies
the caller must also re-derive the tile shape to keep the GM read wide — but today the shape the
user writes is passed straight through to the `tload`, so a natural tiling for `[K, N]` becomes a
pathological one for `[N, K]`. Two directions that would remove the footgun:

1. Let the `tload` widen the read: fill the L1 tile from a K-span wider than one compute chunk and
   serve the chunks from L1, so a shallow-K compute tile no longer forces a shallow-K GM read.
2. Failing that, a `perf_hint` when a weight `tload` under `b_trans` ends up with a contiguous run
   below the fractal size would have surfaced this in seconds — the generated code shows the shape,
   it just is not flagged.

Until then the workaround is entirely on the kernel side and costs nothing structural, so this is
not blocking us; the report is worth keeping open because the shape trap is easy to walk into and
invisible in a `.pto` diff.
